File: services/worker/src/logger/file_logger.py
"""
File Logger with Rotation and PHI Scrubbing

Logging handler that:
- Logs to both console and rotating files
- Scrubs PHI patterns before writing
- Rotates daily and by size
- Keeps 30 days of history

Phase A - Task 38: Logging to Shared Volume with Rotation
"""

import os
import re
import json
import logging
from logging.handlers import RotatingFileHandler, TimedRotatingFileHandler
from datetime import datetime
from typing import Dict, List, Tuple
from pathlib import Path


class PHIScrubFilter(logging.Filter):
    """Filter that scrubs PHI patterns from log messages"""

    # PHI patterns to scrub
    PHI_PATTERNS: List[Tuple[re.Pattern, str]] = [
        (re.compile(r'\b\d{3}-\d{2}-\d{4}\b'), '[SSN_REDACTED]'),
        (re.compile(r'\b\d{3}[-.]?\d{3}[-.]?\d{4}\b'), '[PHONE_REDACTED]'),
        (re.compile(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'), '[EMAIL_REDACTED]'),
        (re.compile(r'\b\d{5}(?:-\d{4})?\b'), '[ZIP_REDACTED]'),
        (re.compile(r'\b(?:MRN|Medical Record|Patient ID)[:\s]+\w+', re.IGNORECASE), '[MRN_REDACTED]'),
        (re.compile(r'\b(?:DOB|Date of Birth)[:\s]+[\d/-]+', re.IGNORECASE), '[DOB_REDACTED]'),
    ]

    def filter(self, record: logging.LogRecord) -> bool:
        """Scrub PHI from log record"""
        # Scrub message
        if isinstance(record.msg, str):
            for pattern, replacement in self.PHI_PATTERNS:
                record.msg = pattern.sub(replacement, record.msg)

        # Scrub args if present
        if record.args:
            scrubbed_args = []
            for arg in record.args:
                if isinstance(arg, str):
                    for pattern, replacement in self.PHI_PATTERNS:
                        arg = pattern.sub(replacement, arg)
                scrubbed_args.append(arg)
            record.args = tuple(scrubbed_args)

        return True


class JSONFormatter(logging.Formatter):
    """Format log records as JSON"""

    def format(self, record: logging.LogRecord) -> str:
        log_data = {
            'timestamp': datetime.utcnow().isoformat(),
            'level': record.levelname,
            'service': 'worker',
            'message': record.getMessage(),
            'module': record.module,
            'function': record.funcName,
            'line': record.lineno,
        }

        # Add exception info if present
        if record.exc_info:
            log_data['exception'] = self.formatException(record.exc_info)

        # Add extra fields
        if hasattr(record, 'extra'):
            log_data.update(record.extra)

        return json.dumps(log_data)


def setup_logger(
    name: str = 'worker',
    log_dir: str = None,
    log_level: str = None,
    max_bytes: int = 100 * 1024 * 1024,  # 100MB
    backup_count: int = 30,
) -> logging.Logger:
    """
    Setup logger with file rotation and PHI scrubbing

    Args:
        name: Logger name
        log_dir: Directory for log files (default: /data/logs)
        log_level: Logging level (default: INFO)
        max_bytes: Max size per log file before rotation
        backup_count: Number of backup files to keep

    Returns:
        Configured logger instance
    """
    # Get configuration from environment
    log_dir = log_dir or os.getenv('LOGS_DIR', '/data/logs')
    log_level = log_level or os.getenv('LOG_LEVEL', 'INFO').upper()

    # Ensure log directory exists
    Path(log_dir).mkdir(parents=True, exist_ok=True)

    # Create logger
    logger = logging.getLogger(name)
    logger.setLevel(getattr(logging, log_level))

    # Remove existing handlers to avoid duplicates
    logger.handlers.clear()

    # Console handler (for container logs)
    console_handler = logging.StreamHandler()
    console_handler.setLevel(getattr(logging, log_level))

    # Use simple format for console
    console_format = logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )
    console_handler.setFormatter(console_format)
    console_handler.addFilter(PHIScrubFilter())
    logger.addHandler(console_handler)

    # File handler with rotation by size
    try:
        log_file = os.path.join(log_dir, f'{name}.log')
        file_handler = RotatingFileHandler(
            log_file,
            maxBytes=max_bytes,
            backupCount=backup_count,
            encoding='utf-8'
        )
        file_handler.setLevel(getattr(logging, log_level))

        # Use JSON format for files
        file_handler.setFormatter(JSONFormatter())
        file_handler.addFilter(PHIScrubFilter())
        logger.addHandler(file_handler)
    except Exception as e:
        logger.error(f"Failed to setup file logging: {e}")

    # Daily rotating file handler
    try:
        daily_log_file = os.path.join(log_dir, f'{name}-daily.log')
        daily_handler = TimedRotatingFileHandler(
            daily_log_file,
            when='midnight',
            interval=1,
            backupCount=backup_count,
            encoding='utf-8'
        )
        daily_handler.setLevel(getattr(logging, log_level))
        daily_handler.setFormatter(JSONFormatter())
        daily_handler.addFilter(PHIScrubFilter())
        logger.addHandler(daily_handler)
    except Exception as e:
        logger.error(f"Failed to setup daily rotating log: {e}")

    return logger


# Create singleton logger instance
logger = setup_logger()


def get_logger(name: str = None) -> logging.Logger:
    """
    Get logger instance

    Args:
        name: Logger name (default: use root worker logger)

    Returns:
        Logger instance
    """
    if name:
        return logging.getLogger(f'worker.{name}')
    return logger

File: services/orchestrator/src/logger/file-logger.ts
/**
 * File Logger with Rotation and PHI Scrubbing
 *
 * Winston-based logger that:
 * - Logs to both console and rotating files
 * - Scrubs PHI patterns before writing
 * - Rotates daily and by size
 * - Keeps 30 days of history
 *
 * Phase A - Task 38: Logging to Shared Volume with Rotation
 */

import * as fs from 'fs';
import * as path from 'path';

// Simple logger interface (can be replaced with Winston later)
interface LogEntry {
  level: string;
  message: string;
  timestamp: string;
  service: string;
  [key: string]: any;
}

class FileLogger {
  private logDir: string;
  private serviceName: string;
  private maxFileSize: number;
  private maxFiles: number;
  private currentDate: string;

  // PHI patterns to scrub
  private phiPatterns = [
    { pattern: /\b\d{3}-\d{2}-\d{4}\b/g, replacement: '[SSN_REDACTED]' },
    { pattern: /\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/g, replacement: '[PHONE_REDACTED]' },
    { pattern: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, replacement: '[EMAIL_REDACTED]' },
    { pattern: /\b\d{5}(?:-\d{4})?\b/g, replacement: '[ZIP_REDACTED]' },
    { pattern: /\b(?:MRN|Medical Record|Patient ID)[:\s]+\w+/gi, replacement: '[MRN_REDACTED]' },
  ];

  constructor(serviceName: string = 'orchestrator') {
    this.serviceName = serviceName;
    this.logDir = process.env.LOGS_DIR || '/data/logs';
    this.maxFileSize = 100 * 1024 * 1024; // 100MB
    this.maxFiles = 30; // Keep 30 days
    this.currentDate = this.getDateString();

    // Ensure log directory exists
    this.ensureLogDir();
  }

  private ensureLogDir() {
    try {
      if (!fs.existsSync(this.logDir)) {
        fs.mkdirSync(this.logDir, { recursive: true });
      }
    } catch (error) {
      console.error('Failed to create log directory:', error);
    }
  }

  private getDateString(): string {
    const now = new Date();
    return now.toISOString().split('T')[0];
  }

  private getLogFilePath(): string {
    return path.join(this.logDir, `${this.serviceName}-${this.currentDate}.log`);
  }

  private scrubPHI(message: string): string {
    let scrubbed = message;
    for (const { pattern, replacement } of this.phiPatterns) {
      scrubbed = scrubbed.replace(pattern, replacement);
    }
    return scrubbed;
  }

  private checkRotation() {
    const today = this.getDateString();
    if (today !== this.currentDate) {
      this.currentDate = today;
    }

    // Check file size
    const filePath = this.getLogFilePath();
    if (fs.existsSync(filePath)) {
      const stats = fs.statSync(filePath);
      if (stats.size > this.maxFileSize) {
        // Rotate by appending timestamp
        const timestamp = Date.now();
        const rotatedPath = path.join(
          this.logDir,
          `${this.serviceName}-${this.currentDate}-${timestamp}.log`
        );
        fs.renameSync(filePath, rotatedPath);
      }
    }

    // Clean old files
    this.cleanOldLogs();
  }

  private cleanOldLogs() {
    try {
      const files = fs.readdirSync(this.logDir);
      const logFiles = files
        .filter(f => f.startsWith(this.serviceName) && f.endsWith('.log'))
        .map(f => ({
          name: f,
          path: path.join(this.logDir, f),
          mtime: fs.statSync(path.join(this.logDir, f)).mtime,
        }))
        .sort((a, b) => b.mtime.getTime() - a.mtime.getTime());

      // Keep only maxFiles newest files
      if (logFiles.length > this.maxFiles) {
        const filesToDelete = logFiles.slice(this.maxFiles);
        filesToDelete.forEach(file => {
          try {
            fs.unlinkSync(file.path);
            console.log(`Deleted old log file: ${file.name}`);
          } catch (error) {
            console.error(`Failed to delete log file ${file.name}:`, error);
          }
        });
      }
    } catch (error) {
      console.error('Failed to clean old logs:', error);
    }
  }

  private writeToFile(entry: LogEntry) {
    try {
      this.checkRotation();

      const logLine = JSON.stringify(entry) + '\n';
      const scrubbedLine = this.scrubPHI(logLine);

      fs.appendFileSync(this.getLogFilePath(), scrubbedLine);
    } catch (error) {
      console.error('Failed to write to log file:', error);
    }
  }

  private log(level: string, message: string, meta?: any) {
    const entry: LogEntry = {
      level,
      message: this.scrubPHI(message),
      timestamp: new Date().toISOString(),
      service: this.serviceName,
      ...meta,
    };

    // Log to console
    const consoleMessage = `[${entry.timestamp}] [${entry.level.toUpperCase()}] ${entry.message}`;
    if (level === 'error') {
      console.error(consoleMessage, meta || '');
    } else if (level === 'warn') {
      console.warn(consoleMessage, meta || '');
    } else {
      console.log(consoleMessage, meta || '');
    }

    // Log to file
    this.writeToFile(entry);
  }

  info(message: string, meta?: any) {
    this.log('info', message, meta);
  }

  warn(message: string, meta?: any) {
    this.log('warn', message, meta);
  }

  error(message: string, meta?: any) {
    this.log('error', message, meta);
  }

  debug(message: string, meta?: any) {
    if (process.env.LOG_LEVEL === 'debug') {
      this.log('debug', message, meta);
    }
  }
}

// Export singleton instance
export const logger = new FileLogger();
